include/mbgl/style/function/type.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <vector>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/variant.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+namespace type {
+
+class Primitive;
+class Variant;
+class Array;
+class NArgs;
+class Typename;
+class Lambda;
+
+using ValueType = variant<
```

Comment:
@jfirebaugh what about something like this instead for that first section ("types of values"):

```c++
class Type {}
class Array : Type {
  Type itemType;
  optional<size_t> length;
}

struct Primitives {
  static Type Null;
  static Type Number;
  static Type String;
  static Type Color;
  static Type Object;
  static Type Value;
}
```